src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectSnapshotWorkerQueue.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class ProjectSnapshotWorkerQueue
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DefaultProjectSnapshotManager _projectManager;
+        private readonly ProjectSnapshotWorker _projectWorker;
+
+        private readonly Dictionary<ProjectId, Project> _projects;
+        private Timer _timer;
+
+        public ProjectSnapshotWorkerQueue(ForegroundDispatcher foregroundDispatcher, DefaultProjectSnapshotManager projectManager, ProjectSnapshotWorker projectWorker)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (projectManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectManager));
+            }
+
+            if (projectWorker == null)
+            {
+                throw new ArgumentNullException(nameof(projectWorker));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _projectManager = projectManager;
+            _projectWorker = projectWorker;
+
+            _projects = new Dictionary<ProjectId, Project>();
+        }
+
+        public bool HasPendingNotifications
+        {
+            get
+            {
+                lock (_projects)
+                {
+                    return _projects.Count > 0;
+                }
+            }
+        }
+
+        public bool IsRunning => _timer != null;
```

Comment:
Not sure if I understood this completely but is null-checking the timer the correct way of checking if it's running? It starts with a delay so not sure if this will be accurate.

Reply:
Yeah, I guess what is running means in the case is `scheduled or running`. It's just for tests.